### ISSUE TITLE
Kế hoạch cho loạt bài viết tiếp theo của Series dl_ap

### ISSUE THREAD

**Issue by gachiemchiep** [dlapplications/dlapplications.github.io#9]
@dl-applications @ducanh841988 @hitogen 

Hiện tại thì series dl_ap đã qua loạt bài về setup công cụ. Loạt bài tiếp theo dự kiến sẽ dùng Tensorflow+Pytorch để implement một  "number plate detection +  recognition" hoàn chỉnh

Dataset dự kiến sử dung:   [LPRdatabase](http://www.medialab.ntua.gr/research/LPRdatabase.html)  hoăc  [ufpr-alpr](https://web.inf.ufpr.br/vri/databases/ufpr-alpr/) (Cần liên hệ với tác giả)

Dự kiến "number plate recognition" sẽ gồm 3 phần

1. Car, bike detection : tìm car, bike trong ảnh (object detection)
2. Number plate detection: tìm region của biển sổ (number plate) từ region của car hoặc bike
3.  Number plate detection: đọc chữ trên biển số 

## Thứ tự bài viết 

Các bài viết sẽ theo thứ tự ngược lại 3, 2, 1

3.  Number plate detection: đọc chữ trên biển số 
Sử dụng các tool đơn giản như thresholding, image segmentation, ... để tách region phần chữ. Sau đấy nhận diện mỗi region và đưa ra kết quả cuối cùng. 
3.1 sử dụng Le-Net + Mnist để train phần nhận diện region                    
3.2  sử dụng Le-Net + dataset khác  để train phần nhận diện region  
3.3  sử dụng Network khác + dataset khác  để train phần nhận diện region  

.2. Number plate detection: tìm region của biển sổ (number plate) từ region của car hoặc bike
Chưa có kế hoạch 

.1. Car, bike detection : tìm car, bike trong ảnh (object detection)
Chưa có kế hoạch 

.0 Tổng hợp 3 phần trên, deploy ra smartphone hoăc phần cứng khác 

Dự kiến mỗi bài sẽ mất 1-2 tuần. Loạt này sẽ kéo dài khoảng hơn 3 tháng. 
Mọi người nghĩ thế nào về kế hoạch này?

**Comment by ducanh841988** [comment]
@gachiemchiep 
phần tìm region có thể sử dụng image processing của openCV. Gồm các bước:
1. Sobel Edge Detection
2. Dung open và close operator để delete nhiễu và kết nối các phần của Region of Interest

**Comment by hitogen** [comment]
@gachiemchiep  phần "number plate recognition" đại ca chia làm 3 với nội dung như vậy nhằm làm chi rứa ạ. Sao phải thay đổi network rồi lại thay đổi dataset? Có đề cập đến transfer learning ko ạ?

**Comment by hitogen** [comment]
@gachiemchiep  @ducanh841988  series deep learning và ứng dụng đang bị đánh số nhầm, các đại ca sửa lại chút nha. Ví dụ bài "Hướng dẫn sử dụng dl" là phần 2, bài tiếp theo "DL Framework" cũng là phần 2, bài tiếp "Cài đặt framework" thì ko có đánh số nữa.

**Comment by hitogen** [comment]
Thêm một vấn đề nữa là khi train CNN thì sẽ cần GPU nên có cần xem xét viết một bài hướng dẫn dùng Google Colab để người đọc có thể xài GPU miễn phí ko ạ?

**Comment by ducanh841988** [comment]
@hitogen Phải xài thử Google Colab thì mới viết tutorial được.

**Comment by nxthuan512** [comment]
@hitogen: Đã thử Colab với GPU theo hướng dẫn (https://medium.com/deep-learning-turkey/google-colab-free-gpu-tutorial-e113627b9f5d) nhưng thấy chậm hơn rất nhiều so với PC đang dùng (không có GPU). Cấu hình PC: CPU Core i7-6700 3.4 GHz, RAM 32-GB (chủ yếu tốn CPU). 

Ngoài ra, Colab không khuyến khích chạy dữ liệu lớn và tốn thời gian nhiều do tài nguyên GPU cần chia sẻ với nhiều users. Code sẽ tự động halt sau 1 thời gian (tầm 15 ph theo thực tế sử dụng)
FAQ: https://research.google.com/colaboratory/faq.html
"**How may I use GPUs and why are they sometimes unavailable?**
Colaboratory is intended for interactive use. Long-running background computations, particularly on GPUs, may be stopped. Please do not use Colaboratory for cryptocurrency mining. Doing so is unsupported and may result in service unavailability. We encourage users who wish to run continuous or long-running computations through Colaboratory’s UI to use a local runtime."

**Comment by hitogen** [comment]
@nxthuan512 thanks đại ca. Vậy là Colab ko khả thi.

**Comment by hitogen** [comment]
@nxthuan512 e thử colab theo cái link đại ca gửi thấy chạy nhanh hơn nhiều so với CPU mà. CPU e là  i5-3210M CPU @ 2.50GHz. Chạy GPU của Colab 1 epoch MNIST hết 10s còn CPU máy cá nhân là gần 4ph cơ ạ.

**Comment by nxthuan512** [comment]
@hitogen A cũng ko rõ. Bên a chạy rất chậm so với CPU. Thử LSTM 400
timesteps, 16 features, dataset 40 MB. Phần preprocessing dataset tốn 15
phút (PC < 1 phút), còn phần training không chạy luôn ...


-----------------------------------------------------------------------------
*Xuan-Thuan NGUYEN*
B.Sc. (Hons), M.Sc., Ph.D.
Postdoctoral Fellow
University of Toronto, Canada
*URL <http://thuanxuannguyen.com>* | *Linkedin
<https://www.linkedin.com/in/xuan-thuan-nguyen-b0287450/> *| *GoogleScholar
<https://scholar.google.co.jp/citations?user=HDiYfIQAAAAJ&hl=en> *|
*ResearchGate
<https://www.researchgate.net/profile/Xuan_Thuan_Nguyen>*
-----------------------------------------------------------------------------

**Comment by gachiemchiep** [comment]
@hitogen 

Để đạt độ chính xác cao nên a chia làm 3 công đoạn.

Ở 3. chia thành các bài khác nhau chủ yếu để người đọc hiểu được cách training. Khi accuracy không đạt như mong muốn thì sẽ lần lượt đổi "network", "dataset", "network+dataset". Như thế sẽ tương ứng với 4 bài ở phần 3.

@nxthuan512 

Mấy cái cloud đấy khi có quá nhiều người dùng sẽ bị bóp băng thông lại, chậm như rùa luôn. Thà dùng Gpu trong máy còn nhanh hơn.

**Comment by hitogen** [comment]
@gachiemchiep có phải ai cũng có GPU trong máy đâu đại ca :))

**Comment by gachiemchiep** [comment]
@dlapplications/dl_ap 

Đã liên lạc được với tác giả và  download xong UFPR-ALPR Dataset

**Comment by hitogen** [comment]
@gachiemchiep  e thấy họ bảo phải gửi từ e-mail trg ĐH mà đại ca down đc hay vậy :)

**Comment by gachiemchiep** [comment]
@hitogen a van giu email cua lab
